### ISSUE TITLE
Make it easy to add load asynchronously qweb2 js templates

### ISSUE THREAD

**Issue by Yajo** [odoo/odoo#20821]
Impacted versions: 9+

Current behavior:

When you want to add a qweb template in JS, you usually do:

```js
qweb.add_template("/path/to/file.xml");
```

This is the common pattern, and it makes the load in sync mode, which blocks the UI. **Bad pattern**

https://github.com/odoo/odoo/blob/053209fb0eeb3f7b6de46ee8887fbf3e2df0b60e/addons/web/static/lib/qweb/qweb2.js#L237-L243

You can load templates asynchronously with this code:

```js
var def = $.Deferred();
qweb.add_template("/path/to/file.xml", function (error, template) {
if (error) {
    def.reject();
} else {
    def.resolve(template);
});
```

There you have then a deferred that resolves when the template is loaded. However, it's hard to use, that's why it isn't the common pattern.

Expected behavior:

qweb should provide a helper, such as `core.qweb.add_template_async()` that returns a deferred object, making it easy to do it the right way.

cc @ged-odoo @Tecnativa

**Comment by Yenthe666** [comment]
Hi @Yajo,

Thanks a lot for bringing up this idea. @ged-odoo I'm quite sure this is interesting for you! What do you think?

**Comment by Yajo** [comment]
I just hit a race condition in https://github.com/OCA/website/pull/402 due to this wrong pattern.

<details>
<summary>Logs:</summary>

Copied from https://travis-ci.org/OCA/website/jobs/329459035#L1723-L1741

```
2018-01-16 13:59:56,923 5966 INFO openerp_test odoo.tests.common: phantomjs: Tour website_form_builder.tour: step body:not(.editor_enable) .s_website_form[data-model_name='res.currency'] input[name=name] succeeded
2018-01-16 13:59:56,930 5966 INFO openerp_test werkzeug: 127.0.0.1 - - [16/Jan/2018 13:59:56] "POST /web/dataset/call HTTP/1.1" 200 -
2018-01-16 13:59:56,950 5966 INFO ? werkzeug: 127.0.0.1 - - [16/Jan/2018 13:59:56] "GET /website_form/static/src/xml/website_form.xml HTTP/1.1" 200 -
2018-01-16 13:59:56,959 5966 INFO openerp_test werkzeug: 127.0.0.1 - - [16/Jan/2018 13:59:56] "POST /website/translations HTTP/1.1" 200 -
2018-01-16 13:59:56,970 5966 INFO openerp_test werkzeug: 127.0.0.1 - - [16/Jan/2018 13:59:56] "GET /website_logo.png HTTP/1.1" 304 -
2018-01-16 13:59:56,993 5966 INFO openerp_test werkzeug: 127.0.0.1 - - [16/Jan/2018 13:59:56] "POST /web/dataset/call_kw/web.planner/search_read HTTP/1.1" 200 -
2018-01-16 13:59:56,996 5966 ERROR openerp_test odoo.tests.common: phantomjs: u"Error: QWeb2: Template 'website_form.status_invalid' not found\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:2908 (in exception)\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:2953 (in _render)\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:2949 (in render)\n at http://127.0.0.1:8069/web/content/558-5c1a45c/web.assets_frontend.js:232 (in update_status)\n at http://127.0.0.1:8069/web/content/558-5c1a45c/web.assets_frontend.js:220 (in send)\n at http://127.0.0.1:8069/web/content/558-5c1a45c/web.assets_frontend.js:219\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:637 (in dispatch)\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:610 (in handle)\n at :0 (in dispatchEvent)\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:3656 (in trigger_mouse_event)\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:3656 (in _click)\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:3655 (in auto)\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:3672\n at :0\n at http://127.0.0.1:8069/web/content/557-98712ed/web.assets_common.js:44\n(leaf frame on top)"
2018-01-16 13:59:57,129 5966 INFO openerp_test odoo.tests.common: phantom_run execution finished
2018-01-16 13:59:57,129 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: FAIL
2018-01-16 13:59:57,136 5966 INFO openerp_test odoo.addons.website_form_builder.tests.test_ui: ======================================================================
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: FAIL: test_ui_website (odoo.addons.website_form_builder.tests.test_ui.UICase)
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: ` Test frontend tour.
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: Traceback (most recent call last):
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: `   File "/home/travis/build/OCA/website/website_form_builder/tests/test_ui.py", line 20, in test_ui_website
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: `     timeout=60)
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: `   File "/home/travis/odoo-10.0/odoo/tests/common.py", line 443, in phantom_js
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: `     self.phantom_run(cmd, timeout)
2018-01-16 13:59:57,136 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: `   File "/home/travis/odoo-10.0/odoo/tests/common.py", line 389, in phantom_run
2018-01-16 13:59:57,137 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: `     "PhantomJS test completed without reporting success; "
2018-01-16 13:59:57,137 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: ` AssertionError: None is not true : PhantomJS test completed without reporting success; the log may contain errors or hints.
2018-01-16 13:59:57,137 5966 INFO openerp_test odoo.addons.website_form_builder.tests.test_ui: Ran 1 test in 18.937s
2018-01-16 13:59:57,137 5966 ERROR openerp_test odoo.addons.website_form_builder.tests.test_ui: FAILED
2018-01-16 13:59:57,137 5966 INFO openerp_test odoo.addons.website_form_builder.tests.test_ui:  (failures=1)
2018-01-16 13:59:57,137 5966 INFO openerp_test odoo.modules.module: odoo.addons.website_form_builder.tests.test_ui tested in 18.94s, 3360 queries
2018-01-16 13:59:57,137 5966 ERROR openerp_test odoo.modules.module: Module website_form_builder: 1 failures, 0 errors
```
</details>

**Comment by ged-odoo** [comment]
@Yajo @Yenthe666 You are right about the fact that doing qweb.add_template("...") is bad and that we need to have a helper to force it to be async.  But it already exists: it is located in ajax.loadXML, and is not really clear looking at its implementation what it does, but just grep for 'ajax.loadXML' and you will find a bunch of calls.  Basically, it takes a path, a qweb and returns a deferred.

However, beware that for some of them (mostly on the website), they are incorrectly used (they are done in the beginning of the module, and the deferred is not kept, so there may have some race conditions when the user clicks too quickly on a button, which triggers the rendering of a template, and will crash if the template is not loaded yet.  We had a few tours randomly failling because of that, and I tried to fix some of them when I could.

So, to summarize:
- we already have the helper you ask for (yay!), so I will close this issue
- we still have a few incorrect uses of ajax.loadXML, we should fix them
- we also should disallow the use of add_template in sync mode, but this is another discussion

Thank you for your feedback, and have a nice day.

**Comment by Yajo** [comment]
It would be really nice to have a `ajax.loadXMLOnce` method, that besides performing the load asynchronously, it just resolves the deferred if the file was already loaded.

This way, when asking for that from a `start: function()` inside an animation, widget, etc., we can safely call it always while not affecting performance.

**Comment by ged-odoo** [comment]
@Yajo Look at this: https://github.com/odoo/odoo/blob/11.0/addons/web/static/src/js/core/ajax.js#L395

The implementation of loadXML is kind of convoluted, but it does what you ask for.  You can call that method as many times as you want, it will cache the deferred on the first call and simply return it for the next ones.  Note that looking at the code, it looks like there is a bug if the same url is called twice with a different qweb engine.  It will actually only add the template to the first qweb, and not to the second one.  Good thing that we do not really use different qweb engines in practice, but still, the whole qweb API could be improved someday

**Comment by ged-odoo** [comment]
In 11.0, we actually often call ajax.loadXML in the willStart method of a widget, to load some templates asynchronously, but before the start (so it can be used as the template rendered by the widget with the template key)

**Comment by Yajo** [comment]
Oh, awesome! I'm hitting problems on v10 addons, eager to migrate them!

Thanks, pal.
